Multi-factor authentication new doc

damienbod opened the pull request:
Fixes https://github.com/dotnet/AspNetCore.Docs/issues/15471

I added rick-anderson as the ms author

@Rick-Anderson @scottaddie @blowdart 

[Internal review page](https://review.docs.microsoft.com/aspnet/core/security/authentication/mfa?view=aspnetcore-3.1&branch=pr-en-us-17112)

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+## Configure MFA for administration pages using ASP.NET Core Identity
+
+MFA could be forced on users to access sensitive pages within an ASP.NET Core Identity application. This could be useful for applications where different levels of access exist for the different identities. For example, users might be able to view the profile data using a password login, but an administrator would be required to use MFA to access the admin pages.
```
```suggestion
MFA could be forced on users to access sensitive pages within an ASP.NET Core Identity app. This could be useful for apps where different levels of access exist for the different identities. For example, users might be able to view the profile data using a password login, but an administrator would be required to use MFA to access the administrative pages.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### Extending the Login with a MFA claim
+
+The demo code is setup using ASP.NET Core with Identity and Razor Pages. The AddIdentity method is used instead of AddDefaultIdentity one, so an IUserClaimsPrincipalFactory implementation can be used to add claims to the identity after a successful login.
```
```suggestion
The demo code is setup using ASP.NET Core with Identity and Razor Pages. The `AddIdentity` method is used instead of `AddDefaultIdentity` one, so an `IUserClaimsPrincipalFactory` implementation can be used to add claims to the identity after a successful login.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+MFA could be forced on users to access sensitive pages within an ASP.NET Core Identity application. This could be useful for applications where different levels of access exist for the different identities. For example, users might be able to view the profile data using a password login, but an administrator would be required to use MFA to access the admin pages.
+
+### Extending the Login with a MFA claim
```
```suggestion
### Extend the login with a MFA claim
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+
+Because we changed the Identity service setup in the Startup class, the layouts of the Identity need to be updated. Scaffold the Identity pages into the application. Define the layout in the Identity/Account/Manage/_Layout.cshtml file.
+
+```razor
````
```suggestion
```cshtml
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+
+Also add the _layout for all the manage pages from the Identity Pages.
+
+```razor
````
```suggestion
```cshtml
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+### Validation the MFA requirement in the Admin Page
````
```suggestion
### Validate the MFA requirement in the admin page
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+
+This policy can then be used in the _Layout view to show or hide the Admin menu with the warning.
+
+```razor
````
```suggestion
```cshtml
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+
+If the identity has logged using MFA, then the Admin menu will be displayed without the warning. If the user has logged without the MFA, the font awesome icon will be displayed, and the tooltip which informs the user, explaining the warning.
+
+```razor
````
```suggestion
```cshtml
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+
+This view will be displayed if the Identity comes from an application which requires MFA but the user has not activated this in Identity. The view informs the user, and adds a link to activate this.
+
+```razor
````
```suggestion
```cshtml
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+
+If the user authenticates without MFA, then the amr claim will probably have a <strong>pwd</strong> value, and the request will not be authorized to access the page. Using the default values, the user will be redirected to the account/AccessDenied page. This can be changed, or you can implement your own custom logic here. In this example, a link is added, so that the valid user can setup MFA for  ...
+
+```razor
````
```suggestion
```cshtml
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+In the ConfigureServices method in the Startup class, the AddOpenIdConnect method is used as the default challenge scheme. The authorization handler which is used to check the amr claim is added as to the IoC. A policy is created then which adds the RequireMfa requirement. 
````
```suggestion
In the `Startup.ConfigureServices` method, the `AddOpenIdConnect` method is used as the default challenge scheme. The authorization handler which is used to check the amr claim is added as to the IoC. A policy is created then which adds the `RequireMfa` requirement.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+The returned value depends on how the identity authenticated and on the Open ID Connect server implementation.
+
+The AuthorizationHandler uses the RequireMfa requirement and validates the **amr** claim. The OpenID Connect server can be implemented using IdentityServer4 with ASP.NET Core Identity. When a user logs in using TOTP, the amr claim is returned with a mfa value. If using a different OpenID Connect server implementation, or a different MFA type, then the amr claim will, or can have a different value ...
```
```suggestion
The `AuthorizationHandler` uses the `RequireMfa` requirement and validates the `amr` claim. The OpenID Connect server can be implemented using IdentityServer4 with ASP.NET Core Identity. When a user logs in using TOTP, the `amr` claim is returned with a MFA value. If using a different OpenID Connect server implementation or a different MFA type, the `amr` claim will, or can, have a different value. The code would need to be extended to accept this as well.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+## Force ASP.NET Core OpenID Connect client to require MFA
+
+In this example it is shown how an ASP.NET Core Razor Page application which uses OpenID Connect to sign in, can require that users have authenticated using Multi-factor authentication. 
```
```suggestion
This example shows how an ASP.NET Core Razor Page app, which uses OpenID Connect to sign in, can require that users have authenticated using MFA.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+In the Login method, the IIdentityServerInteractionService interface implementation _interaction is used to access the Open ID Connnect request parameters. The "acr_values" is accessed using the AcrValues. As the client sent this with **mfa** set, this can then be checked.
+
+If MFA is required, and the user in ASP.NET Core Identity has MFA enabled, then the login continues. If the user has no MFA enabled, the user is redirected to the custom view ErrorEnable2FA.cshtml. Then ASP.NET Core Identity signs the user in.
```
```suggestion
If MFA is required, and the user in ASP.NET Core Identity has MFA enabled, then the login continues. If the user has no MFA enabled, the user is redirected to the custom view *ErrorEnable2FA.cshtml*. Then ASP.NET Core Identity signs the user in.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+<a asp-controller="Manage" asp-action="TwoFactorAuthentication">Enable MFA</a>
+```
+
+In the Login method, the IIdentityServerInteractionService interface implementation _interaction is used to access the Open ID Connnect request parameters. The "acr_values" is accessed using the AcrValues. As the client sent this with **mfa** set, this can then be checked.
````
```suggestion
In the `Login` method, the `IIdentityServerInteractionService` interface implementation _interaction is used to access the Open ID Connnect request parameters. The `acr_values` is accessed using the AcrValues. As the client sent this with **mfa** set, this can then be checked.
```

damienbod replied:
fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### Example OpenID Connect IdentityServer 4 server with ASP.NET Core Identity
+
+On the OpenID Connect server, which is implemented using ASP.NET Core Identity with MVC views, a new view ErrorEnable2FA.cshtml is created, and added.
```
```suggestion
On the OpenID Connect server, which is implemented using ASP.NET Core Identity with MVC views, a new view *ErrorEnable2FA.cshtml* is created and added.
```

damienbod replied:
fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### OpenID Connect ASP.NET Core client
+
+The Razor Page ASP.NET Core Open ID Connnect client application uses the AddOpenIdConnect method to login to the Open ID Connect server. The "acr_values" parameter is set with the "mfa" value and sent with the authentication request. The OpenIdConnectEvents is used to add this.
```
```suggestion
The ASP.NET Core Razor Pages Open ID Connnect client app uses the `AddOpenIdConnect` method to login to the Open ID Connect server. The `acr_values` parameter is set with the `mfa` value and sent with the authentication request. The `OpenIdConnectEvents` is used to add this.
```

damienbod replied:
fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+![Admin MFA authentication](mfa/_static/identitystandalonemfa_01.png)
+
+And when the user clicks the admin link, then the user is redirected to the MFA enable view.
```
```suggestion
When the user clicks the admin link, the user is redirected to the MFA enable view.
```

damienbod replied:
fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+## Send MFA signin requirement to OpenID Connect server 
+
+The **acr_values** parameter can used to pass the mfa required value from the client to the server in an authentication request. 
```
```suggestion
The `acr_values` parameter can used to pass the mfa required value from the client to the server in an authentication request. 
```

damienbod replied:
fixed

damienbod commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+## MFA, 2FA
+
+MFA requires at least two or more types of proof for an identity like something you know, something you possess, or something you inherit for the user to authenticate.
```
fixed

damienbod commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+MFA requires at least two or more types of proof for an identity like something you know, something you possess, or something you inherit for the user to authenticate.
+
+2FA is like a subset of MFA,but the difference being that MFA can require 2 or more factors to prove the identity.
```
fixed

damienbod commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### MFA TOTP (Time-based One-time Password Algorithm)
+
+Multi-factor authentication using TOTP is a supported implementation using ASP.NET Core Identity. This can be used together with the following Apps:
```
fixed

damienbod commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### MFA FIDO2 or Passwordless
+
+FIDO2 is the most secure way of doing Multi-factor authentication and is the only MFA flow which protects against phishing attacks. At present ASP.NET Core does not support FIDO2 directly. FIDO2 can be used for MFA or passwordless flows.
```
fixed

damienbod commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+Azure Active Directory provides support for FIDO2 and passwordless flows.
+
+[Passwordless authentication options for Azure Active Directory](https://docs.microsoft.com/en-us/azure/active-directory/authentication/concept-authentication-passwordless)
```
fixed

damienbod commented:
@scottaddie thanks for the review, added the improvements

damienbod commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### MFA SMS
+
+Although MFA with SMS increases security massively compared with password authentication (Single factor), it is no longer recommended to use SMS as a second factor as too many known attack vectors exist for this type of implementation.
```
fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+MFA using TOTP is a supported implementation using ASP.NET Core Identity. This can be used together with the following apps:
+
+- Microsoft Authenticator App
+- Google Authenticator App
```
```suggestion
* Microsoft Authenticator App
* Google Authenticator App
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### MFA FIDO2 or Passwordless
+
+FIDO2 is the most secure way of achieving MFA and is the only MFA flow which protects against phishing attacks. At present, ASP.NET Core doesn't support FIDO2 directly. FIDO2 can be used for MFA or passwordless flows.
```
```suggestion
FIDO2 is:

* The most secure way of achieving MFA.
* The only MFA flow that protects against phishing attacks.

At present, ASP.NET Core doesn't support FIDO2 directly. FIDO2 can be used for MFA or passwordless flows.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+[Enable QR Code generation for TOTP authenticator apps in ASP.NET Core](xref:security/authentication/identity-enable-qrcodes)
+
+### MFA FIDO2 or Passwordless
```
```suggestion
### MFA FIDO2 or passwordless
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+	{
+		options.AddPolicy("TwoFactorEnabled",
+			x => x.RequireClaim("TwoFactorEnabled", "true" )
+		) ;
```
```suggestion
			x => x.RequireClaim("TwoFactorEnabled", "true")
		);
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+The AdditionalUserClaimsPrincipalFactory adds the **TwoFactorEnabled** claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the claim, ...
````
```suggestion
The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the claim, it would be possible to access the view without MFA directly after activating the MFA.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+            }
+            else
+            {
+                claims.Add(new Claim("TwoFactorEnabled", "false")); ;
```
```suggestion
                claims.Add(new Claim("TwoFactorEnabled", "false"));
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+Because we changed the Identity service setup in the Startup class, the layouts of the Identity need to be updated. Scaffold the Identity pages into the application. Define the layout in the Identity/Account/Manage/_Layout.cshtml file.
````
```suggestion
Because the Identity service setup changed in the `Startup` class, the layouts of the Identity need to be updated. Scaffold the Identity pages into the app. Define the layout in the *Identity/Account/Manage/_Layout.cshtml* file.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+### Validate the MFA requirement in the admin page
+
+The admin Razor Page validates that the user has logged in using MFA. In the OnGet method, the identity is used to access the user claims. The **TwoFactorEnabled** claim is checked for the value true. If the identity is missing this claim, or it is false, the page will redirect to the Enable MFA page. This is possible because the user has logged in already, but without MFA.
```
```suggestion
The administration Razor Page validates that the user has logged in using MFA. In the `OnGet` method, the identity is used to access the user claims. The `TwoFactorEnabled` claim is checked for the value `true`. If the identity is missing this claim or is false, the page redirects to the Enable MFA page. This is possible because the user has logged in already, but without MFA.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+### Validate the MFA requirement in the admin page
````
```suggestion
### Validate the MFA requirement in the administration page
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+@inject IAuthorizationService AuthorizationService
+```
+
+If the identity has logged using MFA, then the Admin menu will be displayed without the warning. If the user has logged without the MFA, the font awesome icon will be displayed, and the tooltip which informs the user, explaining the warning.
````
```suggestion
If the identity has logged in using MFA, the Admin menu is displayed without the warning. If the user has logged in without MFA, the Font Awesome icon is displayed along with the tooltip which informs the user (explaining the warning).
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+The `acr_values` parameter can used to pass the mfa required value from the client to the server in an authentication request. 
+
+> [!NOTE]
+> The **acr_values** parameter needs to be handled on the Open ID Connect server for this to work.
```
```suggestion
> The `acr_values` parameter needs to be handled on the Open ID Connect server for this to work.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+The ASP.NET Core Razor Pages Open ID Connnect client app uses the `AddOpenIdConnect` method to login to the Open ID Connect server. The `acr_values` parameter is set with the `mfa` value and sent with the authentication request. The `OpenIdConnectEvents` is used to add this.
+
+See [Authentication Method Reference Values](https://tools.ietf.org/html/draft-ietf-oauth-amr-values-08) for recommended acr_values values.
```
```suggestion
For recommended `acr_values` parameter values, see [Authentication Method Reference Values](https://tools.ietf.org/html/draft-ietf-oauth-amr-values-08).
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+<a asp-controller="Manage" asp-action="TwoFactorAuthentication">Enable MFA</a>
+```
+
+In the `Login` method, the `IIdentityServerInteractionService` interface implementation _interaction is used to access the Open ID Connnect request parameters. The `acr_values` is accessed using the AcrValues. As the client sent this with **mfa** set, this can then be checked.
````
```suggestion
In the `Login` method, the `IIdentityServerInteractionService` interface implementation _interaction is used to access the Open ID Connnect request parameters. The `acr_values` parameter is accessed using the `AcrValues` property. As the client sent this with `mfa` set, this can then be checked.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+	}
+```
+
+The ExternalLoginCallback works like the local Identity login. The AcrValues property is checked for the "mfa" value and if it is sent, the MFA is forced before the login completes, ie redirected to the ErrorEnable2FA view.
````
```suggestion
The `ExternalLoginCallback` method works like the local Identity login. The `AcrValues` property is checked for the `mfa` value. If the `mfa` value is present, MFA is forced before the login completes (for example, redirected to the `ErrorEnable2FA` view).
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+			false);
+```
+
+If the user is already logged in, the client application still validates the "amr" claim, and can setup the MFA then with a link to the ASP.NET Core Identity view.
````
```suggestion
If the user is already logged in, the client app:

* Still validates the `amr` claim.
* Can set up the MFA with a link to the ASP.NET Core Identity view.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+
+This example shows how an ASP.NET Core Razor Page app, which uses OpenID Connect to sign in, can require that users have authenticated using MFA.
+
+To validate the MFA requirement, an IAuthorizationRequirement requirement is created. This will be added to the pages using a policy which require MFA.
```
```suggestion
To validate the MFA requirement, an `IAuthorizationRequirement` requirement is created. This will be added to the pages using a policy which require MFA.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+An AuthorizationHandler is implemented which will use the **amr** claim and check for the value **mfa**. The amr is returned in the id_token of a successful authentication and can have many different values as defined in the following specification:
````
```suggestion
An `AuthorizationHandler` is implemented that will use the `amr` claim and check for the value `mfa`. The `amr` is returned in the id_token of a successful authentication and can have many different values as defined in the following specification:
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+In the `Startup.ConfigureServices` method, the `AddOpenIdConnect` method is used as the default challenge scheme. The authorization handler which is used to check the amr claim is added as to the IoC. A policy is created then which adds the `RequireMfa` requirement.
````
```suggestion
In the `Startup.ConfigureServices` method, the `AddOpenIdConnect` method is used as the default challenge scheme. The authorization handler, which is used to check the `amr` claim, is added to the Inversion of Control container. A policy is then created which adds the `RequireMfa` requirement.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+This policy is then used in the Razor pages as required. This could be added globally for the whole application as well.
````
```suggestion
This policy is then used in the Razor page as required. The policy could be added globally for the entire app as well.
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+If the user authenticates without MFA, then the amr claim will probably have a <strong>pwd</strong> value, and the request will not be authorized to access the page. Using the default values, the user will be redirected to the account/AccessDenied page. This can be changed, or you can implement your own custom logic here. In this example, a link is added, so that the valid user can setup MFA for  ...
````
```suggestion
If the user authenticates without MFA, the `amr` claim will probably have a `pwd` value. The request won't be authorized to access the page. Using the default values, the user will be redirected to the account/AccessDenied page. This can be changed or you can implement your own custom logic here. In this example, a link is added so that the valid user can set up MFA for his or her account.
```

damienbod commented:
@scottaddie thanks for the review, added your improvements.

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+Also add the _layout for all the manage pages from the Identity Pages.
````
```suggestion
Also assign the layout for all the manage pages from the Identity pages:
```

blowdart replied:
As this touches templates, @HaoK should also review

damienbod replied:
text fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+By [Damien Bowden](https://github.com/damienbod)
+
+Multi-factor authentication (MFA) is a process where a user is requested during a sign-in event for additional forms of identification. This prompt could be to enter a code on their cellphone, use a FIDO2 key or to provide a fingerprint scan. When you require a second form of authentication, security is increased as this additional factor isn't something that's easy for an attacker to obtain or d ...
```
```suggestion
Multi-factor authentication (MFA) is a process where a user is requested during a sign-in event for additional forms of identification. This prompt could be to enter a code on their cellphone, use a FIDO2 key or to provide a fingerprint scan. When you require a second form of authentication, security is increased. The additional factor isn't easily obtained or duplicated by an attacker.
```

blowdart replied:
Enter a code from a cell phone rather than on perhaps?

damienbod replied:
fixed

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the  ...
````
```suggestion
The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims only after a successful login. The claim's value is read from the database. The claim is added here because the user should only access the higher protected view if the identity has logged in with MFA. If the database view is read from the database directly instead of using the claim, it's possible to access the view without MFA directly after activating the MFA.
```

blowdart replied:
Scott this seems to contradict "The claim's value is read from the database" versus "If the database view is read from the database directly instead of using the claim, it's possible to access the view without MFA directly after activating the MFA.". The second statement is correct, the first isn't.

damienbod replied:
text updated

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+### UI logic to show hide information about the user login
````
```suggestion
### UI logic to toggle user login information
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+});
+```
+
+This policy can then be used in the _Layout view to show or hide the Admin menu with the warning.
````
```suggestion
This policy can then be used in the `_Layout` view to show or hide the **Admin** menu with the warning:
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+@inject IAuthorizationService AuthorizationService
+```
+
+If the identity has logged in using MFA, the Admin menu is displayed without the warning. If the user has logged in without MFA, the Font Awesome icon is displayed along with the tooltip which informs the user (explaining the warning).
````
```suggestion
If the identity has logged in using MFA, the **Admin** menu is displayed without the warning. If the user has logged in without MFA, the Font Awesome icon is displayed along with the tooltip that informs the user (explaining the warning).
```

blowdart replied:
Suggest not using fontawesome, we shouldn't be using third party resources on security pages if possible, as if they get compromised ...

damienbod replied:
new text:

If the identity has logged in using MFA, the **Admin** menu is displayed without the warning. If the user has logged in without MFA, an icon is displayed along with the tooltip that informs the user (explaining the warning).

scottaddie replied:
I suggest we move away from icons and use text instead.

damienbod replied:
done

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+## Send MFA signin requirement to OpenID Connect server 
+
+The `acr_values` parameter can used to pass the mfa required value from the client to the server in an authentication request. 
```
```suggestion
The `acr_values` parameter can used to pass the `mfa` required value from the client to the server in an authentication request. 
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+<a asp-controller="Manage" asp-action="TwoFactorAuthentication">Enable MFA</a>
+```
+
+In the `Login` method, the `IIdentityServerInteractionService` interface implementation _interaction is used to access the Open ID Connnect request parameters. The `acr_values` parameter is accessed using the `AcrValues` property. As the client sent this with `mfa` set, this can then be checked.
````
```suggestion
In the `Login` method, the `IIdentityServerInteractionService` interface implementation `_interaction` is used to access the Open ID Connect request parameters. The `acr_values` parameter is accessed using the `AcrValues` property. As the client sent this with `mfa` set, this can then be checked.
```

blowdart replied:
"OpenID Connect"

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+An `AuthorizationHandler` is implemented that will use the `amr` claim and check for the value `mfa`. The `amr` is returned in the id_token of a successful authentication and can have many different values as defined in the following specification:
````
```suggestion
An `AuthorizationHandler` is implemented that will use the `amr` claim and check for the value `mfa`. The `amr` is returned in the `id_token` of a successful authentication and can have many different values as defined in the following specification:
```

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+If the user authenticates without MFA, the `amr` claim will probably have a `pwd` value. The request won't be authorized to access the page. Using the default values, the user will be redirected to the account/AccessDenied page. This can be changed or you can implement your own custom logic here. In this example, a link is added so that the valid user can set up MFA for his or her account.
````
```suggestion
If the user authenticates without MFA, the `amr` claim will probably have a `pwd` value. The request won't be authorized to access the page. Using the default values, the user will be redirected to the *account/AccessDenied* page. This can be changed or you can implement your own custom logic here. In this example, a link is added so that the valid user can set up MFA for his or her account.
```

blowdart replied:
Ah there's the special claim :)

damienbod replied:
fixed the texted here

blowdart commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+### MFA FIDO2 or passwordless
+
+FIDO2 is:
```
is currently?

blowdart commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+### MFA TOTP (Time-based One-time Password Algorithm)
+
+MFA using TOTP is a supported implementation using ASP.NET Core Identity. This can be used together with the following apps:
```
with any compliant authenticator application, including

damienbod replied:
fixed

blowdart commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+## MFA, 2FA
+
+MFA requires at least two or more types of proof for an identity like something you know, possess, or inherit for the user to authenticate.
```
What's meant by inherit here?

damienbod replied:
Updated text:

MFA requires at least two or more types of proof for an identity like something you know, something you possess, or biometric validation for the user to authenticate.

blowdart commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+Azure Active Directory provides support for FIDO2 and passwordless flows. For more information, see [Passwordless authentication options for Azure Active Directory](/azure/active-directory/authentication/concept-authentication-passwordless).
+
+You can implement ASP.NET Core with FIDO2 by using the following OSS FIDO2 implementation:
```
Given this is a security topic should we link to things we haven't examined for security?

damienbod replied:
removed the OSS link

blowdart commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,562 @@
+
+### MFA SMS
+
+MFA with SMS increases security massively compared with password authentication (Single factor). However, using SMS as a second factor is no longer recommended. Too many known attack vectors exist for this type of implementation.
```
Link to NIST guidelines here? https://pages.nist.gov/800-63-3/sp800-63b.html

damienbod replied:
added the link

blowdart commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the  ...
````
There's actually a standard claim for this, which azure ad and others send. Identity doesn't unfortunately.

HaoK replied:
Should we file and issue for identity for this @blowdart seems like a cheapie

blowdart replied:
I thought we had to be honest, Dominick wanted it

damienbod replied:
now using standard claim amr with values mfa, pwd

damienbod commented:
@scottaddie @blowdart 

I updated using the review feedback. I left some conversations opened, because I'm not sure if I fixed this correctly, or you meant something else.

Could you add some feedback or close the review conversation if this is ok

thanks Damien

scottaddie commented on aspnetcore/security/authentication/mfa.md:
```diff
@@ -0,0 +1,560 @@
+			   data-toggle="tooltip" 
+			   data-placement="bottom" 
+			   title="MFA is NOT enabled. This is required for the Admin Page. If you have activated MFA, then logout, login again.">
+				<i class="fa fa-exclamation-triangle" aria-hidden="true"></i>
```
Per Barry's recommendation, remove references to Font Awesome (including these CSS class names). It might make more sense to use text instead of an icon.

damienbod replied:
@scottaddie Removed, thanks

scottaddie commented on aspnetcore/security/authentication/mfa.md:
````diff
@@ -0,0 +1,559 @@
+@inject IAuthorizationService AuthorizationService
+```
+
+If the identity has logged in using MFA, the **Admin** menu is displayed without the tooltip warning. If the user has logged in without MFA, an icon is displayed along with the tooltip that informs the user (explaining the warning).
````
This "icon" text needs to be updated now that Font Awesome has been removed.

damienbod replied:
updated to: the **Admin (Not Enabled)** menu is displayed

scottaddie commented:
Ready for @blowdart to take another look.

damienbod commented:
@blowdart Anything else we need to do here? Thanks for you help

Greetings Damien

scottaddie commented:
Paging @blowdart for a final review.

scottaddie commented:
@damienbod Thank you for your patience and hard work on this PR! It's very much appreciated.
